### ISSUE TITLE
scrape (histograms): Investigate (and address) protobuf scraping performance problems

### ISSUE THREAD

**Issue by beorn7** [prometheus/prometheus#14668]
### Proposal

tl;dr: To scrape native histograms, the scraping has to happen using protobuf. The current protobuf parsing creates a lot of memory churn. In practice, that mostly boils down to increased CPU usage for garbage collection. Depending on the exact usage pattern of a Prometheus server, the CPU usage may grow significantly (we have seen a 2x increase in real-world use-cases). This can hit users hard if they "simply" switch on native histograms, and it will be especially surprising if they only want to ingest a few native histograms (but via activating the feature flag, they switch _all_ scraping to protobuf).

__An important implication for declaring native histograms a stable feature: We cannot simply "switch on native histograms" by default if that implies that all scrapes are using protobuf scraping while protobuf scraping uses so much more CPU as we have seen now.__ (Or in other words: This issue needs to get resolved to declare native histograms a stable feature.)

### Various pieces of context

- Prometheus currently uses the (unmaintained) gogo-protobuf library. It already has a much better performance than the official Go protobuf library. Independent from this issue, we have plans (and need) to migrate away from gogo-protobuf to an actually maintained library. It is preferred to get this done soon, and to start the performance optimizations with the new library to avoid redoing it twice (before and after migrating).
- To get to the ground truth, we need to do proper profiling first. Having said that, I assume that the source of the problem is that most expositions are still dominated strongly by strings (metric names, label names, label values), and the text parser is highly optimized to efficiently ingest those strings, while the generated protobuf parsing code is creating separate Go strings for all those strings in the protobuf before our own label handling code can even kick in.
- The current parsing and scrape code is highly intertwined without much abstraction barriers, utilizing specifics of the text format to achieve high ingestion performance. For example, if the part of the line that defines the metric (name, labels) hasn't changed compared to the last scrape, there is no parsing happening and the code just re-uses the already known label set. The protobuf parsing has to work against this approach and essentially simulate a line-based layout from parsing the protobuf exposition. For one, the protobuf parsing (including allocating all the strings) is happening upon each scrape over and over again, and on top of that, the parsing code has to emulate a "line based view" of the exposition so that it can be ingested properly.

### Immediate work-around

In a scenario where only selected few endpoints expose native histograms, users can switch back to text based scraping on endpoints where they don't expect native histograms, using the [`scrape_protocols` setting in the scrape config](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config).

### Work-around in the not too far future

Once native histograms can be scraped via the text format, we can make that the default (or users can switch to it). Even though heavy native histograms might be less efficient to scrape in the text format, most expositions will still be dominated by conventional metrics so that overall, the performance impact should be much more manageable than that of a complete switch to protobuf for everything.

### Actual solutions

In the midterm, we should anyway refactor the scrape code to not be "hardwired to the text format", which might already give us a performance gain for protobuf (while still aiming for not losing much performance when scraping the text format).

In general, we need to address the bottlenecks as identified by profiling. In the ideal case, we can optimize things just by changing the way we use our protobuf library of choice. Less ideal (and more likely) is the case where we need a more invasive approach to avoid allocating many strings for each scrape over and over again. The most extreme approach would be to hand-code the entire protobuf decoding.

**Comment by bboreham** [comment]
I started looking at this; my idea is that we can use the protobuf-encoded LabelPairs as the key to the scrape cache.

**Comment by bwplotka** [comment]
Some pre-optimization benchmark, built on top of @bboreham work in https://github.com/prometheus/prometheus/pull/15720

```
benchstat -col /fmt scrape-loop-v1.txt    
goos: darwin
goarch: arm64
pkg: github.com/prometheus/prometheus/scrape
cpu: Apple M1 Pro
                   │  PromText   │              OMText               │             PromProto              │
                   │   sec/op    │   sec/op     vs base              │   sec/op     vs base               │
ScrapeLoopAppend-2   32.45µ ± 8%   30.68µ ± 2%  -5.46% (p=0.002 n=6)   54.48µ ± 8%  +67.88% (p=0.002 n=6)

                   │   PromText   │            OMText             │              PromProto               │
                   │     B/op     │     B/op      vs base         │     B/op      vs base                │
ScrapeLoopAppend-2   20.30Ki ± 9%   19.74Ki ± 3%  ~ (p=0.394 n=6)   75.58Ki ± 1%  +272.28% (p=0.002 n=6)

                   │  PromText  │              OMText              │              PromProto               │
                   │ allocs/op  │ allocs/op   vs base              │  allocs/op   vs base                 │
ScrapeLoopAppend-2   12.00 ± 0%   13.00 ± 0%  +8.33% (p=0.002 n=6)   810.00 ± 0%  +6650.00% (p=0.002 n=6)
```

Profiles: [mem](https://pprof.me/5997398552bcc21ac2a883b94573816b) [cpu](https://pprof.me/423a60c3d6a114a860cd888b5705f629)

The main problem is allocating those strings for labels when unmarshalling each delimited message (usually per metric family?).

![image](https://github.com/user-attachments/assets/c3f31cf6-7b2f-40a9-9723-20b832063635)

Honestly, the best solution would be something like [Thanos ZLabels here](https://github.com/thanos-io/thanos/blob/6f03fcb0ba777d4f6159b5ec502c8ac0ab6d927b/pkg/store/labelpb/label.go#L134), yoloString-ing those as we build labels.Labels from this anyway later on. Good question how to best inject this without gogo proto goodies, not sure. After that Opaque API will help, but only if we pool things really (e.g. Metrics), otherwise just OpaqueAPI might not help as we use all the fields in general.

**Comment by bboreham** [comment]
> as we build labels.Labels from this anyway later on

This is only true the first time you see one particular series; after that the Labels are held by the scrape cache.
What we need is a token that can be used to look up the scrape cache with a minimum of allocation.

I don't see why that shouldn't be the relevant subsection from the protobuf stream, including all tags and length bytes.
This might be helped by the opaque API; I haven't investigated closely.

**Comment by bwplotka** [comment]
Got it, we need to fix that caching key part in order to reuse strings 👍🏽

**Comment by bboreham** [comment]
If we can find a token to cache from then we don't need to worry about the strings.

**Comment by bwplotka** [comment]
After 3 attempts I found something promising. Looks scary, but it's the only way with the long term to work with vtprotobuf one some streaming plugin. 

https://github.com/prometheus/prometheus/pull/15731

**Comment by bwplotka** [comment]
FYI: More realistic benchmarks shows the current (before optimizations) code is actually still faster than text parsers, but allocates more. For all gauges, in a single metric family we we see slow down. Extended the benchmark to showcase this: https://github.com/prometheus/prometheus/pull/15966

With my optimizations it looks very promising for realistic case (3x faster than text parses, 10% less allocated memory) and equal resources for the just gauges. 🤞🏽 Trying to finish that up https://github.com/prometheus/prometheus/pull/15731

**Comment by bwplotka** [comment]
Some PRs to review for this: https://cloud-native.slack.com/archives/C01AUBA4PFE/p1739367866185279
